internal/validators/rbac.go

```diff
@@ -100,46 +103,70 @@ func (s *RBACRuleService) processPermissionSet(set v1alpha1.PermissionSet, princ
 		}
 	}
 
-	// First, find out whether we need to look the role up by its role name if the user provided
-	// its role name instead of its name.
-	var roleName string
-	role := set.Role
-	if role.Name != nil {
-		roleName = *role.Name
-	} else if role.RoleName != nil {
-		// To do the role name lookup, we need to get all of the role definitions that exist in the
-		// subscription that we're working with. We figure out which subscription we're working with
-		// by using the subscription from the scope of the permission set we're working on.
-		subForLookup, err := azure_utils.RoleAssignmentScopeSubscription(set.Scope)
-		if err != nil {
-			s.log.V(0).Error(err, "failed to parse subscription ID from scope string to perform role name lookup")
-			return err
+	roleName := set.Role
+
+	_, ok := foundRoleNames[roleName]
+	if !ok {
+		*failures = append(*failures, fmt.Sprintf("Principal missing role %s", roleName))
+	}
+
+	if err := s.processRolePermissions(set, failures); err != nil {
+		return fmt.Errorf("failed to process permissions specified in permission set: %w", err)
+	}
+
+	return nil
+}
+
+// processRolePermissions processes the permissions specified in a permissions set. This means it
+// verifies that the permissions specified in the spec are indeed present in the role definition for
+// the role specified. This is only used when permissions are specified. Otherwise, role is assumed
+// to have all the needed permissions.
+func (s *RBACRuleService) processRolePermissions(set v1alpha1.PermissionSet, failures *[]string) error {
+
+	// Special case. Nothing to do, so skip Azure API call and don't append any failures.
+	if len(set.Actions) == 0 && len(set.DataActions) == 0 {
+		return nil
+	}
+
+	// Get all the permissions that a particular role definition has. There are two ways to get
+	// the details of a role definition. You can use its fully-qualified ID or a combination of its
+	// id and its scope. We do the latter because this is information that will always be available
+	// in our spec.
+	perms, err := s.rdAPI.GetPermissionDataForRoleDefinition(set.Role, set.Scope)
+	if err != nil {
+		return fmt.Errorf("failed to get permission data for role definition: %w", err)
+	}
+
+	// Permission data is divided into "actions" and "data actions". We need to deal with both, but
+	// actions don't relate to data actions and vice versa.
+	if len(set.Actions) > 0 {
+		if string_utils.AnyNil(perms.Actions) || string_utils.AnyNil(perms.NotActions) {
```

Comment:
:pick: You're incurring additional overhead to iterate over both these slices first during the nil check, then again inside `ToVals`. Why not just have `ToVals` return an error if a nil is encountered? Then `ToVals` is nicer too, since it doesn't make any assumptions.